lib/undiciRetry.ts

```diff
@@ -0,0 +1,112 @@
+import type { Client, Dispatcher } from 'undici'
+import type { IncomingHttpHeaders } from 'undici/types/header'
+import type { Either } from './either'
+import { ResponseError } from './ResponseError'
+import { setTimeout } from 'node:timers/promises'
+import { errors } from 'undici'
+
+const TIMEOUT_ERRORS = [errors.BodyTimeoutError.name, errors.HeadersTimeoutError.name]
+
+export type RequestResult<T> = {
+  body: T
+  headers: IncomingHttpHeaders
+  statusCode: number
+}
+
+export type RetryConfig = {
+  maxAttempts: number
+  delayBetweenAttemptsInMsecs: number
+  statusCodesToRetry: readonly number[]
+  retryOnTimeout: boolean
+  safeParseJson?: boolean
+}
+
+const DEFAULT_RETRY_CONFIG: RetryConfig = {
+  maxAttempts: 1,
+  delayBetweenAttemptsInMsecs: 0,
+  statusCodesToRetry: [],
+  retryOnTimeout: false,
+  safeParseJson: false,
+}
+
+export async function sendWithRetry<T>(
+  client: Client,
+  request: Dispatcher.RequestOptions,
+  retryConfig: RetryConfig = DEFAULT_RETRY_CONFIG
+): Promise<Either<RequestResult<unknown>, RequestResult<T>>> {
+  let attemptsSoFar = 0
+
+  while (true) {
+    attemptsSoFar++
+    try {
+      const response = await client.request(request)
+
+      // success
+      if (response.statusCode < 400) {
+        const resolvedBody = await resolveBody(response, retryConfig.safeParseJson)
+        return {
+          result: {
+            body: resolvedBody,
+            headers: response.headers,
+            statusCode: response.statusCode,
+          },
+        }
+      }
+
+      // Do not retry, return last error response
+      if (
+        retryConfig.statusCodesToRetry.indexOf(response.statusCode) === -1 ||
+        attemptsSoFar >= retryConfig.maxAttempts
+      ) {
+        const resolvedBody = await resolveBody(response)
+        return {
+          error: {
+            body: resolvedBody,
+            headers: response.headers,
+            statusCode: response.statusCode,
+          },
+        }
+      }
+
+      // retry
+      // undici response body always has to be processed or discarded
+      await response.body.dump()
+    } catch (err: any) {
+      // on internal client error we can't do much; if there are still retries left, we retry, if not, we rethrow an error
```

Comment:
`ResponseError` thrown by `resolveBody()` might get swallowed here. But perhaps that's alright?